mllib/src/main/scala/org/apache/spark/ml/feature/Bucketizer.scala

```diff
@@ -0,0 +1,158 @@
+import org.apache.spark.ml.util.SchemaUtils
+import org.apache.spark.ml.{Estimator, Model}
+import org.apache.spark.sql._
+import org.apache.spark.sql.functions._
+import org.apache.spark.sql.types.{DoubleType, StructField, StructType}
+
+/**
+ * :: AlphaComponent ::
+ * `Bucketizer` maps a column of continuous features to a column of feature buckets.
+ */
+@AlphaComponent
+private[ml] final class Bucketizer(override val parent: Estimator[Bucketizer])
+  extends Model[Bucketizer] with HasInputCol with HasOutputCol {
+
+  def this() = this(null)
+
+  /**
+   * Parameter for mapping continuous features into buckets. With n splits, there are n+1 buckets.
+   * A bucket defined by splits x,y holds values in the range [x,y).
+   * @group param
+   */
+  val splits: Param[Array[Double]] = new Param[Array[Double]](this, "splits",
+    "Split points for mapping continuous features into buckets. With n splits, there are n+1" +
+      " buckets. A bucket defined by splits x,y holds values in the range [x,y).",
+    Bucketizer.checkSplits)
+
+  /** @group getParam */
+  def getSplits: Array[Double] = $(splits)
+
+  /** @group setParam */
+  def setSplits(value: Array[Double]): this.type = set(splits, value)
+
+  /** @group Param */
+  val lowerInclusive: BooleanParam = new BooleanParam(this, "lowerInclusive",
+    "An indicator of the inclusiveness of negative infinite.")
+  setDefault(lowerInclusive -> true)
+
+  /** @group getParam */
+  def getLowerInclusive: Boolean = $(lowerInclusive)
+
+  /** @group setParam */
+  def setLowerInclusive(value: Boolean): this.type = set(lowerInclusive, value)
+
+  /** @group Param */
+  val upperInclusive: BooleanParam = new BooleanParam(this, "upperInclusive",
+    "An indicator of the inclusiveness of positive infinite.")
+  setDefault(upperInclusive -> true)
+
+  /** @group getParam */
+  def getUpperInclusive: Boolean = $(upperInclusive)
+
+  /** @group setParam */
+  def setUpperInclusive(value: Boolean): this.type = set(upperInclusive, value)
+
+  /** @group setParam */
+  def setInputCol(value: String): this.type = set(inputCol, value)
+
+  /** @group setParam */
+  def setOutputCol(value: String): this.type = set(outputCol, value)
+
+  override def transform(dataset: DataFrame): DataFrame = {
+    transformSchema(dataset.schema)
+    val wrappedSplits = Array(Double.MinValue) ++ $(splits) ++ Array(Double.MaxValue)
+    val bucketizer = udf { feature: Double =>
+      Bucketizer
+        .binarySearchForBuckets(wrappedSplits, feature, $(lowerInclusive), $(upperInclusive)) }
+    val newCol = bucketizer(dataset($(inputCol)))
+    val newField = prepOutputField(dataset.schema)
+    dataset.withColumn($(outputCol), newCol.as($(outputCol), newField.metadata))
+  }
+
+  private def prepOutputField(schema: StructType): StructField = {
+    val attr = new NominalAttribute(
+      name = Some($(outputCol)),
+      isOrdinal = Some(true),
+      values = Some($(splits).map(_.toString)))
```

Comment:
Each value should correspond to 1 bin, not 1 split, right?  This will have to take the inclusive parameters into account too.